shared/src/pool_cache.rs

```diff
@@ -130,116 +181,118 @@ impl PoolCache {
             }
             tokio::time::sleep(self.delay_between_retries).await;
         }
         fetch().await
     }
-}
 
-#[async_trait::async_trait]
-impl PoolFetching for PoolCache {
-    async fn fetch(&self, pairs: HashSet<TokenPair>, block: Block) -> Result<Vec<Pool>> {
+    pub async fn fetch(&self, keys: impl IntoIterator<Item = K>, block: Block) -> Result<Vec<V>> {
         let block = match block {
             Block::Recent => None,
             Block::Number(number) => Some(number),
         };
 
+        let mut cache_hit_count = 0usize;
         let mut cache_hits = Vec::new();
         let mut cache_misses = HashSet::new();
         let last_update_block;
         {
             let mut mutexed = self.mutexed.lock().unwrap();
-            for &pair in &pairs {
-                match mutexed.get(pair, block) {
-                    Some(pools) => cache_hits.extend_from_slice(&pools),
+            for key in keys {
+                match mutexed.get(key.clone(), block) {
+                    Some(values) => {
+                        cache_hit_count += 1;
+                        cache_hits.extend_from_slice(values);
+                    }
                     None => {
-                        cache_misses.insert(pair);
+                        cache_misses.insert(key);
                     }
                 }
             }
             last_update_block = mutexed.last_update_block;
         }
 
         self.metrics
-            .pools_fetched(pairs.len() - cache_misses.len(), cache_misses.len());
+            .entries_fetched(cache_hit_count, cache_misses.len());
 
         if cache_misses.is_empty() {
             return Ok(cache_hits);
         }
 
         let cache_miss_block = block.unwrap_or(last_update_block);
-        let uncached_pools = self
+        let uncached_values = self
             .fetch_inner(cache_misses.clone(), Block::Number(cache_miss_block))
             .await?;
+
+        cache_hits.extend_from_slice(&uncached_values);
+
         {
             let mut mutexed = self.mutexed.lock().unwrap();
-            mutexed.insert(cache_miss_block, cache_misses.into_iter(), &uncached_pools);
+            mutexed.insert(cache_miss_block, cache_misses.into_iter(), uncached_values);
         }
 
-        cache_hits.extend_from_slice(&uncached_pools);
         Ok(cache_hits)
     }
 }
 
-#[async_trait::async_trait]
-impl Maintaining for PoolCache {
-    async fn run_maintenance(&self) -> Result<()> {
-        let block = current_block::block_number(&self.block_stream.borrow())?;
-        self.update_cache(block)
-            .await
-            .context("failed to update pool cache")
-    }
-}
-
 #[derive(Debug)]
-struct Mutexed {
-    recently_used: LruCache<TokenPair, ()>,
-    // For quickly finding at which block a pair is cached.
-    cached_most_recently_at_block: HashMap<TokenPair, u64>,
+struct Mutexed<K, V>
+where
+    K: CacheKey<V>,
+{
+    recently_used: LruCache<K, ()>,
```

Comment:
I guess `LruCache<K, ()>` is like a map with only keys and the values don't matter.

Reply:
Correct, I think its because there is no `LruSet<T>`.